cg/services/order_validation_service/workflows/tomte/validation/inter_field/pedigree/models.py

```diff
@@ -0,0 +1,83 @@
+from cg.services.order_validation_service.models.errors import (
+    DescendantAsFatherError,
+    DescendantAsMotherError,
+    PedigreeError,
+    SampleIsOwnFatherError,
+    SampleIsOwnMotherError,
+)
+from cg.services.order_validation_service.workflows.tomte.models.case import TomteCase
+from cg.services.order_validation_service.workflows.tomte.models.sample import TomteSample
+
+
+class Node:
+    def __init__(self, sample: TomteSample):
+        self.sample = sample
+        self.father: Node | None = None
+        self.mother: Node | None = None
+        self.visited = False
+        self.in_stack = False
+
+
+class Pedigree:
+    def __init__(self, case: TomteCase):
+        self.graph: dict[str, Node] = {}
+        self.case = case
+        self._add_nodes()
+        self._add_parents()
+
+    def _add_nodes(self) -> None:
+        for sample in self.case.samples:
+            node = Node(sample=sample)
+            self.graph[sample.name] = node
+
+    def _add_parents(self) -> None:
+        for node in self.graph.values():
+            sample: TomteSample = node.sample
+            if sample.mother:
+                node.mother = self.graph.get(sample.mother)
+            if sample.father:
+                node.father = self.graph.get(sample.father)
+
+    @property
+    def nodes(self) -> list[Node]:
+        return self.graph.values()
+
+    def validate(self) -> list[PedigreeError]:
+        errors = []
+        for node in self.nodes:
+            if not node.visited:
+                self._detect_cycles(node, errors)
+        return errors
+
+    def _detect_cycles(self, node: Node, errors: list[PedigreeError]) -> None:
+        node.visited = True
+        node.in_stack = True
+
+        parents = {"mother": node.mother, "father": node.father}
+
+        for parent_type, parent in parents.items():
+            if parent and parent.in_stack:
+                error = self._get_error(node=node, parent_type=parent_type)
+                errors.append(error)
+            elif parent and not parent.visited:
+                self._detect_cycles(parent, errors)
+
+        node.in_stack = False
```

Comment:
Changed the name to `in_current_path`. The DFS traverses the graph until it can reach further and then backtracks. During the backtrack, it is necessary to mark that the node is no longer in the current path!